### PR TITLE
Add HTTP path rewriting from protobuf annotations

### PR DESCRIPTION
**Pull request by brummer-simon**
Add rewriting of HTTP request path with protobuf annotations if given.

### CODE REVIEW

File: source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/filter.cc [hunk 1]
```diff
@@ -95,44 +97,63 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
...
+    if (!status_method.ok()) {
+      ENVOY_LOG(error, "Failed to parse HTTP Method. Abort Processing. Error was: {}",
+                status_method.status().message());
+
+      respondWithGrpcError(*decoder_callbacks_, Errors::UnexpectedMethodType);
+      return Http::FilterHeadersStatus::StopIteration;
+    }
+
+    auto path = static_cast<std::string>(headers.getPathValue());
+    auto method_and_path = HttpMethodAndPath{*status_method, path};
```
Comment (brummer-simon) [inline source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/filter.cc]: Note to self: path could be moved here

File: source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/http_methods.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Convert HttpMethod enum value to a string.
+ * @param[in] method the http method to convert into a string.
+ * @returns The string representation of @p method.
+ */
+std::string httpMethodToString(HttpMethod method);
+
+/**
+ * Type typing a http method and a path together
```
Comment (brummer-simon) [inline source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/http_methods.h]: tying not typing

File: source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/transcoder.cc [hunk 3]
```diff
@@ -49,8 +50,9 @@ class Transcoder::Impl : public Logger::Loggable<Logger::Id::filter> {
 private:
   DescriptorPoolPtr descriptors_;
   TypeResolverPtr type_resolver_;
-  MethodInfoMap method_infos_;
-  MethodName method_name_;
+  MethodInfoMap grpc_method_infos_;
+  MethodInfo const* selected_grpc_method_;
+  HttpMethodAndPath selected_html_method_and_path_;
```
Comment (brummer-simon) [inline source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/transcoder.cc]: Rename to http path. HTML is not right here.

File: source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/filter.cc [hunk 4]
```diff
@@ -95,44 +97,63 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
...
+    if (!status_method.ok()) {
+      ENVOY_LOG(error, "Failed to parse HTTP Method. Abort Processing. Error was: {}",
+                status_method.status().message());
+
+      respondWithGrpcError(*decoder_callbacks_, Errors::UnexpectedMethodType);
+      return Http::FilterHeadersStatus::StopIteration;
+    }
+
+    auto path = static_cast<std::string>(headers.getPathValue());
+    auto method_and_path = HttpMethodAndPath{*status_method, std::move(path)};
```
Comment (aledzu) [inline source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/filter.cc:122]: Is status_metod string? Can we move it as well?

File: source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/filter.cc [hunk 5]
```diff
@@ -95,44 +97,63 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
...
+    auto status_transcoder = transcoder_.prepareTranscoding(method_and_path);
+    if (!status_transcoder.ok()) {
+      ENVOY_LOG(error, "Failed to prepare transcoding. Abort Processing. Error was: {}",
+                status_transcoder.message());
 
       // TODO: Add fitting error grpc error
       respondWithGrpcError(*decoder_callbacks_, Errors::GrpcFrameTooSmall);
```
Comment (aledzu) [inline source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/filter.cc]: Add the new type of error.

File: source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/filter.cc [hunk 6]
```diff
@@ -95,44 +97,63 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
...
+    // Rewrite HTTP Headers
+    auto status_path = transcoder_.getHttpRequestPath();
+    if (!status_path.ok()) {
+      ENVOY_LOG(error, "Failed to determine new HTTP Request path. Abort Processing. Error was: {}",
+                status_path.status().message());
+
+      respondWithGrpcError(*decoder_callbacks_, Errors::UnexpectedRequestPath);
+      return Http::FilterHeadersStatus::StopIteration;
+    }
+    headers.setPath(*status_path);
```
Comment (aledzu) [inline source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/filter.cc]: move?

File: source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/http_methods.cc [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  static const std::map<std::string, HttpMethod> mapping{
+      {"GET", HttpMethod::GET},
+      {"POST", HttpMethod::POST},
+      {"PUT", HttpMethod::PUT},
+      {"DELETE", HttpMethod::DELETE},
+  };
+
+  std::string normalized{maybe_http_method};
+  std::transform(normalized.begin(), normalized.end(), normalized.begin(),
+                 [](auto ch) { return std::toupper(ch); });
```
Comment (aledzu) [inline source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/http_methods.cc:19]: Is this safe to do? Write at the beginning of the container that is under transformation?

File: source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/http_methods.cc [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+  auto pos = mapping.find(normalized);
```
Comment (aledzu) [inline source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/http_methods.cc]: const

File: source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/transcoder.cc [hunk 9]
```diff
@@ -102,38 +104,84 @@ absl::Status Transcoder::Impl::init(Api::Api& api, std::string const& proto_desc
...
 bool Transcoder::Impl::isInitialized() const { return descriptors_ && type_resolver_; }
 
-absl::Status Transcoder::Impl::prepareTranscoding(std::string const& method_name) {
+absl::Status Transcoder::Impl::prepareTranscoding(HttpMethodAndPath method_and_path) {
   assert(isInitialized());
 
-  auto stripped_method_name = method_name.substr(method_name.find_last_of('/') + 1);
-  if (method_infos_.find(stripped_method_name) == method_infos_.cend()) {
-    return absl::NotFoundError(absl::StrCat("Failed to find method name: ", method_name));
+  auto& path = method_and_path.path;
```
Comment (aledzu) [inline source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/transcoder.cc]: add const where applicable.

File: source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/filter.cc (hunk 4 above)
Comment (brummer-simon) [reply source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/filter.cc:122]: Nope its an enum value so move would be a copy and is cheap anyway.

File: source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/http_methods.cc (hunk 7 above)
Comment (brummer-simon) [reply source/extensions/filters/http/grpc_http1_reverse_bridge_transcoder/http_methods.cc:19]: acording to the stl documentation, it is. It is basically an inplace transformation

**Review by aledzu — APPROVED** [review]

**Pull request merged**